read_test.go

```diff
@@ -595,6 +595,103 @@ func TestReadPixelData_SkipPixelData(t *testing.T) {
+func headerWithNoFileMetaInformationGroupLength() (*bytes.Buffer, error) {
+	var buf bytes.Buffer
+
+	dcm_writer := NewWriter(&buf)
+	dcm_writer.SetTransferSyntax(binary.LittleEndian, true)
+
+	elements := []*Element{
+		mustNewElement(tag.MediaStorageSOPClassUID, []string{"SecondaryCapture"}),
+		mustNewElement(tag.MediaStorageSOPInstanceUID, []string{"1.3.6.1.4.1.35190.4.1.20210608.607733549593"}),
+		mustNewElement(tag.TransferSyntaxUID, []string{"=RLELossless"}),
+		mustNewElement(tag.ImplementationClassUID, []string{"1.6.6.1.4.1.9590.100.1.0.100.4.0"}),
+		mustNewElement(tag.SOPInstanceUID, []string{"1.3.6.1.4.1.35190.4.1.20210608.607733549593"}),
+	}
+	for _, e := range elements {
+		err := dcm_writer.WriteElement(e)
+		if err != nil {
+			return nil, err
+		}
+	}
+	data := buf.Bytes()
+	// Construct valid DICOM header preamble
+	data = append([]byte("DICM"), data...)
+	preamble := make([]byte, 128)
+	data = append(preamble, data...)
+	dcm_header := bytes.NewBuffer(data)
+	return dcm_header, nil
+}
+
+// Returns a fake DICOM group 2 header with a FileMetaInformationGroupLength tag (0x0002,0x0000)
+func headerWithFileMetaInformationGroupLength() (*bytes.Buffer, error) {
+	var buf bytes.Buffer
+
+	dcm_writer := NewWriter(&buf)
+	dcm_writer.SetTransferSyntax(binary.LittleEndian, true)
+
+	elements := []*Element{
+		mustNewElement(tag.FileMetaInformationGroupLength, []int{188}),
+		mustNewElement(tag.FileMetaInformationVersion, []byte{0x00, 0x01}),
+		mustNewElement(tag.MediaStorageSOPClassUID, []string{"1.2.276.0.7230010.3.1.0.1"}),
+		mustNewElement(tag.MediaStorageSOPInstanceUID, []string{"1.3.6.1.4.1.35190.4.1.20210608.607733549593"}),
+		mustNewElement(tag.TransferSyntaxUID, []string{"=RLELossless"}),
+		mustNewElement(tag.ImplementationClassUID, []string{"1.2.276.0.7230010.3.0.3.6.7"}),
+		mustNewElement(tag.ImplementationVersionName, []string{"OFFIS_DCMTK_367"}),
+		mustNewElement(tag.SOPInstanceUID, []string{"1.3.6.1.4.1.35190.4.1.20210608.607733549593"}),
+	}
+	for _, e := range elements {
+		err := dcm_writer.WriteElement(e)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Construct valid DICOM header preamble
+	data := buf.Bytes()
+	data = append([]byte("DICM"), data...)
+	preamble := make([]byte, 128)
+	data = append(preamble, data...)
+	dcm_header := bytes.NewBuffer(data)
+	return dcm_header, nil
+}
+
+func TestReadHeader_TryAllowErrorMetaElementGroupLength(t *testing.T) {
+	opts := parseOptSet{allowErrorMetaElementGroupLength: true}
+
+	t.Run("NoFileMetaInformationGroupLength", func(t *testing.T) {
+		dcmheader_noinfogrplen, err := headerWithNoFileMetaInformationGroupLength()
+		if err != nil {
+			t.Errorf("unsuccesful generation of fake header data")
+		} else {
+			r := &reader{
+				rawReader: dicomio.NewReader(bufio.NewReader(dcmheader_noinfogrplen), binary.LittleEndian, int64(dcmheader_noinfogrplen.Len())),
+				opts:      opts,
+			}
+			_, err = r.readHeader()
```

Comment:
good job! note the advantage we have of bootstrapping the elements in this test ourselves (vs. in parse_test, where we are reading dicoms from disk) is that we should be able to easily compare what was parsed in readHeader to the `[]*Element` that we expect to be there, since those were setup by the test. If you have the data setup function return the `[]*Element` expected as well, you should be able to use `cmp` to compare them. You'll have to use a special version, something like:

https://github.com/suyashkumar/dicom/blob/48ba291715cc85f31c5133f81fd9d5d5dc720d2c/parse_internal_test.go#L45-L47

Reply:
I heavily rewrote the header generation utility functions and updated the tests to include the compare of the written elements vs the ones  read with readHeader.